apps/pipelines/nodes/base.py

```diff
@@ -176,6 +189,43 @@ def get_all_routes(self) -> dict:
 
         return routes_dict
 
+    def get_node_output_by_name(self, node_name: str) -> Any:
+        """
+        Get the output of a node by its name.
+        """
+        output = self["outputs"].get(node_name)
+        if output:
+            return output["message"]
+        return None
+
+    def get_node_output(self, node_id: str) -> Any:
+        """
+        Get the output of a node by its ID.
+        """
+        for output in self["outputs"].values():
+            if output.get("node_id") == node_id:
+                return output["message"]
+        return None
+
+    def get_node_inputs(self, node_id: str, incoming_nodes: list[str]) -> dict[str, Any]:
+        """
+        Get the inputs for the given node based on the node's incoming edges.
+
+        Returns:
+            A dictionary mapping incoming node IDs to their respective outputs in the state.
+            If there is no output for an incoming edge or if that edge was not targeted,
+            the value in the output will be None.
+        """
+        inputs = {}
+        for incoming_node_id in incoming_nodes:
+            targets = [step[2] for step in self["path"] if step[1] == incoming_node_id]
```

Comment:
I'm having a hard time keeping track of what the `step` indexes mean. How about we use constants to indicate what each index refers to?